src/Runtime/OMSort.inc

```diff
@@ -0,0 +1,453 @@
+  const uint64_t rank = omTensorGetRank(inputTensor);
+  assert(rank <= 6 && "omTensorSort assumes rank <= 6");
+  assert(axis == (rank - 1) && "omTensorSort assumes axis == (rank - 1)");
+  const int64_t *inputShape = omTensorGetShape(inputTensor);
+  const int64_t *inputStrides = omTensorGetStrides(inputTensor);
+  assert(inputStrides[axis] == 1 && "omTensorSort assumes strides[axis] == 1");
+  void *orderPtr = omTensorGetDataPtr(orderTensor);
+  uint64_t *order = (uint64_t *)orderPtr;
+  void *dataPtr = omTensorGetDataPtr(inputTensor);
+  int64_t sort_elems = inputShape[axis];
+  // Sorting not necessary for empty array
+  if (sort_elems == 0)
+    return;
+
+  // Choose the appropriate compare function
+  compareFunctionType *compareElements =
+      getCompareFunction(ascending, dataType);
+  uint64_t datasize = getDataSize(dataType);
+
+  // Choose the sorting function
+  sortFunctionType *sortFunc = bubble_sort;
+  switch (algorithm) {
+  case OMSORT_BUBBLESORT:
+    sortFunc = bubble_sort;
+    break;
+  case OMSORT_COMBSORT:
+    sortFunc = comb_sort;
+    break;
+  case OMSORT_QUICKSORT:
+#if defined(__APPLE__)
+    // MacOS supports qsort_r in different API
+    sortFunc = qsort_r;
+#elif defined(_MSC_VER)
+    // Windows supports qsort_s
+    sortFunc = qsort_s;
+#else
+    // Use standard quick sort in libc
+    sortFunc = qsort_r;
+#endif
+    break;
+  case OMSORT_QUICKSORT_CUSTOM:
+    sortFunc = quick_sort_custom; // custum quick sort
+    break;
+  default:
+    assert("invalid sorting algorithm");
+    break;
+  }
+  // To support input Tensor with various ranks in a uniform way.
+  // If the input rank < 6, upgrade the rank to 6 virtually without changing
+  // the physical memory layout by inserting length=1 ranks at lower ranks.
+  // The 5th axis becomes the sort axis.
+  int64_t shape[6] = {1, 1, 1, 1, 1, 1};
+  int64_t strides[6] = {0, 0, 0, 0, 0, 0};
+  for (uint64_t i = 0; i < rank; i++) {
+    shape[i + (6 - rank)] = inputShape[i];
+    strides[i + (6 - rank)] = inputStrides[i];
+  }
+
+  // Sort the the 6th axis in the outer 5 loops
+  for (int dim0 = 0; dim0 < shape[0]; dim0++) {
+    for (int dim1 = 0; dim1 < shape[1]; dim1++) {
+      for (int dim2 = 0; dim2 < shape[2]; dim2++) {
+        for (int dim3 = 0; dim3 < shape[3]; dim3++) {
+          for (int dim4 = 0; dim4 < shape[4]; dim4++) {
+            uint64_t off = dim0 * strides[0] + dim1 * strides[1] +
+                           dim2 * strides[2] + dim3 * strides[3] +
+                           dim4 * strides[4];
+            void *data = ((char *)dataPtr) + datasize * off;
+            uint64_t *idx = order + off;
+#if defined(__APPLE__)
+            sortFunc((void *)idx, sort_elems, sizeof(uint64_t), data,
+                compareElements);
+#else
+            sortFunc((void *)idx, sort_elems, sizeof(uint64_t), compareElements,
+                data);
+#endif
+          }
+        }
+      }
+    }
+  }
+  return;
+}
```

Comment:
The implementation of qsort seems way too complicated than necessary. Maybe this is something you can make use of for `OMSort.inc`:
```
#include <assert.h>
#include <stddef.h>

#ifndef __cplusplus
#define __USE_GNU
#include <stdlib.h>
#endif

#include "onnx-mlir/Runtime/OnnxDataType.h"

/* Compare function for case when there is an index array and a data array,
 * and the array being sorted is the index array but value comparison is
 * using the data array.
 */
#define COMPARE(type)                                                          \
  do {                                                                         \
    int i1 = *((const type *)p1);                                              \
    int i2 = *((const type *)p2);                                              \
    return (((const type *)qarg->data_ptr)[i1] >                               \
               ((const type *)qarg->data_ptr)[i2]) -                           \
           (((const type *)qarg->data_ptr)[i1] <                               \
               ((const type *)qarg->data_ptr)[i2]);                            \
  } while (0)

/* Custom data passed into the qsort_r/s function */
typedef struct qsort_arg {
  void *data_ptr;         /* pointer to data array  */
  OM_DATA_TYPE data_type; /* ONNX data type         */
} qsort_arg_t;

/* Use comp_ascending for ascending sort,
 * use comp_descending for descending sort.
 */
#if defined(__APPLE__) || defined(_MSC_VER)

extern int comp_ascending(void *arg, const void *p1, const void *p2);
extern int compdecending(void *arg, const void *p1, const void *p2);

int comp_decending(void *arg, const void *p1, const void *p2) {
  return comp_ascending(arg, p2, p1);
}

int comp_ascending(void *arg, const void *p1, const void *p2) {

#else /* Linux */

extern int comp_ascending(const void *p1, const void *p2, void *arg);
extern int comp_decending(const void *p1, const void *p2, void *arg);

int comp_decending(const void *p1, const void *p2, void *arg) {
  return comp_ascending(p2, p1, arg);
}

int comp_ascending(const void *p1, const void *p2, void *arg) {

#endif

  qsort_arg_t *qarg = (qsort_arg_t *)arg;

  switch (qarg->data_type) {
  case ONNX_TYPE_BOOL:
    COMPARE(bool);
  case ONNX_TYPE_UINT8:
    COMPARE(uint8_t);
  case ONNX_TYPE_INT8:
    COMPARE(int8_t);
  case ONNX_TYPE_UINT16:
    COMPARE(uint16_t);
  case ONNX_TYPE_INT16:
    COMPARE(int16_t);
  case ONNX_TYPE_UINT32:
    COMPARE(uint32_t);
  case ONNX_TYPE_INT32:
    COMPARE(int32_t);
  case ONNX_TYPE_UINT64:
    COMPARE(uint64_t);
  case ONNX_TYPE_INT64:
    COMPARE(int64_t);
  case ONNX_TYPE_FLOAT:
    COMPARE(float);
  case ONNX_TYPE_DOUBLE:
    COMPARE(double);
  default:
    assert(false && "unexpected data type in compare function");
  }
}

/* quick sort wrapper function to deal with slightly different qsort
 * variations on different platforms.
 *
 * For sorting an index array whose values are in a separate data array,
 * use arg to pass in a pointer to qsort_arg_t struct.
 *
 * For ascending sort, use the compare function comp_ascending.
 * For descending sort, use the compare function comp_descending.
 */
void quick_sort(void *base, size_t num, size_t size,
#if defined(__APPLE__) || defined(_MSC_VER)
    int (*comp)(void *, const void *, const void *),
#else /* Linux */
    int (*comp)(const void *, const void *, void *),
#endif
    void *arg) {
#if defined(__APPLE__)
  qsort_r(base, num, size, arg, comp);

#elif defined(_MSC_VER)
  qsort_s(base, num, size, comp, arg);

#else /* Linux */
  qsort_r(base, num, size, comp, arg);
#endif
}
```

Reply:
Thank you for the comment. I agree that your method is simpler.

Let me explain the reason why type-specific sort and compare functions are used.

As background, the sort and compare functions are called in very tight loop so many times (80 * 10646 * 10647/2 times in the Yolov3 case), and only the sort function is called in the loop as you can see the code below.

Type specific sort and compare functions are used for the performance optimization, by using type-specific sort and compare functions, we can move "switch" statements to check the data type called in the loop to the outsize of the loop. And I suppose that it improves the performance.

```
  // Sort the the 6th axis in the outer 5 loops
  for (int dim0 = 0; dim0 < shape[0]; dim0++) {
    for (int dim1 = 0; dim1 < shape[1]; dim1++) {
      for (int dim2 = 0; dim2 < shape[2]; dim2++) {
        for (int dim3 = 0; dim3 < shape[3]; dim3++) {
          for (int dim4 = 0; dim4 < shape[4]; dim4++) {
            uint64_t off = dim0 * strides[0] + dim1 * strides[1] +
                           dim2 * strides[2] + dim3 * strides[3] +
                           dim4 * strides[4];
            void *data = ((char *)dataPtr) + datasize * off;
            uint64_t *idx = order + off;
#if defined(__APPLE__)
            sortFunc((void *)idx, sort_elems, sizeof(uint64_t), data, compareElements);
#else
            sortFunc((void *)idx, sort_elems, sizeof(uint64_t), compareElements, data);
#endif
          }
        }
      }
    }
  }
```

Reply:
OK thanks for the background. It would be nice to have some comments about all these complexities are for optimizing performance.

Reply:
I appended some descriptions to explain the reason why data-specific functions are used. Thank you for the comments.